url7f729337e3_timmolter_Sundial_tgz-pJ8-org_knowm_sundial_SampleRunJ8/src/org/quartz/triggers/CronExpression.java

```diff
@@ -0,0 +1,1571 @@
+    expressionParsed = true;
+
+    try {
+
+      if (seconds == null) {
+        seconds = new TreeSet<Integer>();
+      }
+      if (minutes == null) {
+        minutes = new TreeSet<Integer>();
+      }
+      if (hours == null) {
+        hours = new TreeSet<Integer>();
+      }
+      if (daysOfMonth == null) {
+        daysOfMonth = new TreeSet<Integer>();
+      }
+      if (months == null) {
+        months = new TreeSet<Integer>();
+      }
+      if (daysOfWeek == null) {
+        daysOfWeek = new TreeSet<Integer>();
+      }
+      if (years == null) {
+        years = new TreeSet<Integer>();
+      }
+
+      int exprOn = SECOND;
+
+      StringTokenizer exprsTok = new StringTokenizer(expression, " \t", false);
+
+      while (exprsTok.hasMoreTokens() && exprOn <= YEAR) {
+        String expr = exprsTok.nextToken().trim();
+
+        // throw an exception if L is used with other days of the month
+        if (exprOn == DAY_OF_MONTH && expr.indexOf('L') != -1 && expr.length() > 1 && expr.contains(",")) {
+          throw new ParseException("Support for specifying 'L' and 'LW' with other days of the month is not implemented", -1);
+        }
+        // throw an exception if L is used with other days of the week
+        if (exprOn == DAY_OF_WEEK && expr.indexOf('L') != -1 && expr.length() > 1 && expr.contains(",")) {
+          throw new ParseException("Support for specifying 'L' with other days of the week is not implemented", -1);
+        }
+        if (exprOn == DAY_OF_WEEK && expr.indexOf('#') != -1 && expr.indexOf('#', expr.indexOf('#') + 1) != -1) {
+          throw new ParseException("Support for specifying multiple \"nth\" days is not implemented.", -1);
+        }
+
+        StringTokenizer vTok = new StringTokenizer(expr, ",");
+        while (vTok.hasMoreTokens()) {
+          String v = vTok.nextToken();
+          storeExpressionVals(0, v, exprOn);
+        }
+
+        exprOn++;
+      }
+
+      if (exprOn <= DAY_OF_WEEK) {
+        throw new ParseException("Unexpected end of expression.", expression.length());
+      }
+
+      if (exprOn <= YEAR) {
+        storeExpressionVals(0, "*", YEAR);
+      }
+
+      TreeSet<Integer> dow = getSet(DAY_OF_WEEK);
+      TreeSet<Integer> dom = getSet(DAY_OF_MONTH);
+
+      // Copying the logic from the UnsupportedOperationException below
+      boolean dayOfMSpec = !dom.contains(NO_SPEC);
+      boolean dayOfWSpec = !dow.contains(NO_SPEC);
+
+      if (!dayOfMSpec || dayOfWSpec) {
+        if (!dayOfWSpec || dayOfMSpec) {
+          throw new ParseException("Support for specifying both a day-of-week AND a day-of-month parameter is not implemented.", 0);
+        }
+      }
+    } catch (ParseException pe) {
+      throw pe;
+    } catch (Exception e) {
+      throw new ParseException("Illegal cron expression format (" + e.toString() + ")", 0);
+    }
+  }
+
+  protected int storeExpressionVals(int pos, String s, int type) throws ParseException {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 66. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.